### ISSUE TITLE
--enable-split-objs doesn't work with modules containing apostrophes

### ISSUE THREAD

**Issue by RyanGlScott** [haskell/cabal#3631]
Originally reported here: https://github.com/NixOS/nixpkgs/issues/15916#issuecomment-234406671

This manifests when trying to install `generic-deriving-1.10.6` with `--enable-split-objs` on. I've managed to reproduce this two versions of `ld`. Here is with the `ld` from `binutils-2.25.1`:

```
$ cabal install generic-deriving --enable-split-objs -j1
Resolving dependencies...
Configuring generic-deriving-1.10.6...
Building generic-deriving-1.10.6...
Preprocessing library generic-deriving-1.10.6...
[ 1 of 19] Compiling Paths_generic_deriving ( dist/dist-sandbox-3194207/build/autogen/Paths_generic_deriving.hs, dist/dist-sandbox-3194207/build/Paths_generic_deriving.o )
[ 2 of 19] Compiling Generics.Deriving.TH.Internal ( src/Generics/Deriving/TH/Internal.hs, dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Internal.o )
[ 3 of 19] Compiling Generics.Deriving.TH.Pre4'9 ( src/Generics/Deriving/TH/Pre4'9.hs, dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4'9.o )
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __1.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __2.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __3.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __4.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __5.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __6.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __7.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __8.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __9.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __10.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __11.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __12.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __13.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __14.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __15.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __16.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __17.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __18.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __19.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __20.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __21.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __22.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __23.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __24.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __25.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __26.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __27.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __28.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __29.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __30.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __31.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __32.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __33.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __34.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __35.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __36.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __37.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __38.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __39.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __40.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __41.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __42.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __43.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __44.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __45.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __46.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __47.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __48.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __49.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __50.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __51.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __52.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __53.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __54.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __55.o
/usr/bin/ld: cannot find /home/ryanglscott/Documents/Hacking/Haskell/sandbox/generic-deriving-1.10.6/dist/dist-sandbox-3194207/build/Generics/Deriving/TH/Pre4 inside 
/usr/bin/ld: cannot find _o_split/Pre4
/usr/bin/ld: cannot find __56.o
collect2: error: ld returned 1 exit status
'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9'9
Failed to install generic-deriving-1.10.6
cabal: Error: some packages failed to install:
generic-deriving-1.10.6 failed during the building phase. The exception was:
ExitFailure 1
$ ld --version
GNU ld (GNU Binutils for Ubuntu) 2.25.1
Copyright (C) 2014 Free Software Foundation, Inc.
This program is free software; you may redistribute it under the terms of
the GNU General Public License version 3 or (at your option) a later version.
This program has absolutely no warranty.
```

And with the `ld` from `binutils-2.26.1`:

```
$ cabal install generic-deriving --enable-split-objs -j1
Resolving dependencies...
Configuring generic-deriving-1.10.6...
Building generic-deriving-1.10.6...
Preprocessing library generic-deriving-1.10.6...
[ 1 of 19] Compiling Paths_generic_deriving ( dist/build/autogen/Paths_generic_deriving.hs, dist/build/Paths_generic_deriving.o )
[ 2 of 19] Compiling Generics.Deriving.TH.Internal ( src/Generics/Deriving/TH/Internal.hs, dist/build/Generics/Deriving/TH/Internal.o )
[ 3 of 19] Compiling Generics.Deriving.TH.Pre4'9 ( src/Generics/Deriving/TH/Pre4'9.hs, dist/build/Generics/Deriving/TH/Pre4'9.o )
collect2: fatal error: ld terminated with signal 11 [Segmentation fault], core dumped
compilation terminated.
/usr/bin/ld: error: /tmp/ghc1563_0/ghc_38.ldscript:1:87: invalid character
/usr/bin/ld: error: /tmp/ghc1563_0/ghc_38.ldscript:1:87: syntax error, unexpected $end, expecting '-' or STRING or QUOTED_STRING or AS_NEEDED
/usr/bin/ld: error: /tmp/ghc1563_0/ghc_38.ldscript: not an object or archive
Failed to install generic-deriving-1.10.6
cabal: Error: some packages failed to install:
generic-deriving-1.10.6 failed during the building phase. The exception was:
ExitFailure 1
$ ld --version
GNU gold (GNU Binutils for Ubuntu 2.26.1) 1.11
Copyright (C) 2015 Free Software Foundation, Inc.
This program is free software; you may redistribute it under the terms of
the GNU General Public License version 3 or (at your option) a later version.
This program has absolutely no warranty.
```

**Comment by 23Skidoo** [comment]
So we need to do some mangling somewhere in the pipeline. Not sure whether this a Cabal or GHC bug.

**Comment by Ptival** [comment]
Are you sure it's not an issue of just escaping a command for `ld`?

**Comment by 23Skidoo** [comment]
Looks like a GHC bug:

```
$ ghc --make Main\'Main.hs 
[1 of 1] Compiling Main             ( Main'Main.hs, Main'Main.o )
Linking Main'Main ...
$ ghc --make -split-objs Main\'Main.hs
[1 of 1] Compiling Main             ( Main'Main.hs, Main'Main.o )
/usr/bin/ld: cannot find /home/refold/tttt/Main
/usr/bin/ld: cannot find Main_o_split/Main
/usr/bin/ld: cannot find Main__1.o
/usr/bin/ld: cannot find /home/refold/tttt/Main
/usr/bin/ld: cannot find Main_o_split/Main
/usr/bin/ld: cannot find Main__2.o
/usr/bin/ld: cannot find /home/refold/tttt/Main
/usr/bin/ld: cannot find Main_o_split/Main
/usr/bin/ld: cannot find Main__3.o
/usr/bin/ld: cannot find /home/refold/tttt/Main
/usr/bin/ld: cannot find Main_o_split/Main
/usr/bin/ld: cannot find Main__4.o
collect2: error: ld returned 1 exit status
''''''''
```

Please [file an issue on the GHC bug tracker](https://ghc.haskell.org/trac/ghc/wiki/ReportABug).

**Comment by RyanGlScott** [comment]
Ah, now that I search through the GHC Trac tickets, it looks like this was fixed in GHC 8.0 via [#11063](https://ghc.haskell.org/trac/ghc/ticket/11063), in which the fix was, as @Ptival guessed, to quote file paths in linker scripts. Indeed, I cannot reproduce the issue with GHC 8.0.1.

Sorry for the noise.

**Comment by 23Skidoo** [comment]
np!
